internal/policy/container/has_modified_files.go

```diff
@@ -180,69 +202,242 @@ func (p HasModifiedFilesCheck) Metadata() check.Metadata {
 	}
 }
 
-// Untar takes a destination path and a reader; a tar reader loops over the tarfile
-// creating the file structure at 'dst' along the way, and writing any files
-func untar(pathChan chan<- string, r io.Reader) error {
-	tr := tar.NewReader(r)
+func extractPackageNameVersionRelease(pkgList []*rpmdb.PackageInfo) map[string]struct{} {
+	pkgNameList := make(map[string]struct{}, len(pkgList))
+	for _, pkg := range pkgList {
+		pkgNameList[fmt.Sprintf("%s-%s-%s", pkg.Name, pkg.Version, pkg.Release)] = struct{}{}
+	}
+	return pkgNameList
+}
 
-	for {
-		header, err := tr.Next()
+// FindRPMDB attempts to extract a valid RPMDB from layers in the order
+// they are provided. If found is not set to true, foundIndex and pkglist should
+// be disregarded as any value there will be invalid.
+func findRPMDB(ctx context.Context, layer v1.Layer) (found bool, pkglist []*rpmdb.PackageInfo) {
+	logger := logr.FromContextOrDiscard(ctx)
+	var err error
+	pkglist, err = extractRPMDB(ctx, layer)
+	if err == nil {
+		id, _ := layer.Digest()
+		logger.V(log.TRC).Info("findRPMDB found an RPM db", "layer", id)
+		found = true
+		return
+	}
 
-		switch {
-		// if no more files are found return
-		case err == io.EOF:
-			close(pathChan)
-			return nil
+	return found, pkglist
+}
 
-		// return any other error
-		case err != nil:
-			return err
+// DirectoryIsExcluded excludes a directory and any file contained in that directory.
+func directoryIsExcluded(ctx context.Context, s string) bool {
+	excl := map[string]struct{}{
+		"etc": {},
+		"var": {},
+		"run": {},
+	}
 
-		// if the header is nil, just skip it (not sure how this happens)
-		case header == nil:
-			continue
+	for k := range excl {
+		if strings.HasPrefix(s, filepath.Clean(k+"/")) || k == s {
+			logger := logr.FromContextOrDiscard(ctx)
+			logger.V(log.TRC).Info("directory excluded", "directory", s)
+			return true
 		}
+	}
 
-		// the target location where the dir/file should be created
-		target := header.Name
+	return false
+}
 
-		// the following switch could also be done using fi.Mode(), not sure if there
-		// a benefit of using one vs. the other.
-		// fi := header.FileInfo()
+// PathIsExcluded checks if s is excluded explicitly as written.
+func pathIsExcluded(ctx context.Context, s string) bool {
+	excl := map[string]struct{}{
```

Comment:
Do we need `var` and `/var` here as well?

Reply:
Var is caught in the directoryIsExcluded